crates/bevy_core_pipeline/src/dof/dof.wgsl

```diff
@@ -0,0 +1,275 @@
+// efficient than performing a 2D convolution.
+//
+// The Bokeh blur uses a similar, but more complex, separable convolution
+// technique. The algorithm is described in Colin Barré-Brisebois, "Hexagonal
+// Bokeh Blur Revisited" [1]. It's motivated by the observation that we can use
+// separable convolutions not only to produce boxes but to produce
+// parallelograms. Thus, by performing three separable convolutions in sequence,
+// we can produce a hexagonal shape. The first and second convolutions are done
+// simultaneously using multiple render targets to cut the total number of
+// passes down to two.
+//
+// [1]: https://colinbarrebrisebois.com/2017/04/18/hexagonal-bokeh-blur-revisited-part-2-improved-2-pass-version/
+
+#import bevy_core_pipeline::fullscreen_vertex_shader::FullscreenVertexOutput
+#import bevy_pbr::view_transformations::depth_ndc_to_view_z
+#import bevy_render::view::View
+
+// Parameters that control the depth of field effect. See
+// `bevy_core_pipeline::dof::DepthOfFieldUniforms` for information on what these
+// parameters mean.
+struct DepthOfFieldParams {
+    /// The distance in meters to the location in focus.
+    focal_distance: f32,
+
+    /// The [focal length]. Physically speaking, this represents "the distance
+    /// from the center of the lens to the principal foci of the lens". The
+    /// default value, 50 mm, is considered representative of human eyesight.
+    /// Real-world lenses range from anywhere from 5 mm for "fisheye" lenses to
+    /// 2000 mm for "super-telephoto" lenses designed for very distant objects.
+    ///
+    /// The higher the value, the more blurry objects not in focus will be.
+    ///
+    /// [focal length]: https://en.wikipedia.org/wiki/Focal_length
+    focal_length: f32,
+
+    /// The premultiplied factor that we scale the circle of confusion by.
+    ///
+    /// This is calculated as `focal_length² / (sensor_height * aperture_f_stops)`.
+    coc_scale_factor: f32,
+
+    /// The maximum diameter, in pixels, that we allow a circle of confusion to be.
+    ///
+    /// A circle of confusion essentially describes the size of a blur.
+    ///
+    /// This value is nonphysical but is useful for avoiding pathologically-slow
+    /// behavior.
+    max_circle_of_confusion_diameter: f32,
+    max_depth: f32,
+}
+
+// The first bokeh pass outputs to two render targets. We declare them here.
+struct DualOutput {
+    // The vertical output.
+    @location(0) output_0: vec4<f32>,
+    // The diagonal output.
+    @location(1) output_1: vec4<f32>,
+}
+
+// The view uniforms, containing the camera matrices.
+//
+// These may look unused, but they're actually used by some of the helper
+// functions (e.g. `depth_ndc_to_view_z`).
+@group(0) @binding(0) var<uniform> view: View;
+
+// The depth texture for the main view. This nee
+#ifdef MULTISAMPLED
+@group(0) @binding(1) var depth_texture: texture_depth_multisampled_2d;
```

Comment:
This doesn't seem to work on webgl2 (both chrome and firefox panic with `Tex storage 2D multisample is not supported`).

I'm not sure what wgpu downlevel flag or feature we need to check for to avoid having this panic.

Reply:
Yuck. We might have to disable depth of field with MSAA on on those platforms. I don't see a way around it unfortunately.

Reply:
It's disabled.